temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -180,149 +243,375 @@ public void handle(LocalActivityTask task) throws Exception {
+
+      if (RetryOptionsUtils.areAttemptsReached(retryOptions, timingOutAttempt)) {
+        executionContext.callback(
+            failed(
+                activityId,
+                RetryState.RETRY_STATE_MAXIMUM_ATTEMPTS_REACHED,
+                newTimeoutFailure(
+                    TimeoutType.TIMEOUT_TYPE_START_TO_CLOSE, executionContext.getLastFailure()),
+                timingOutAttempt,
+                null));
+        return;
       }
 
-      RetryOptions retryOptions = buildRetryOptions(activityTask.getRetryPolicy());
-
-      long sleepMillis = retryOptions.calculateSleepTime(attempt);
-      long elapsedTask = System.currentTimeMillis() - activityStartTimeMs;
-      long sinceScheduled =
-          System.currentTimeMillis() - Timestamps.toMillis(activityTask.getScheduledTime());
-      long elapsedTotal = elapsedTask + sinceScheduled;
-      Duration timeout = ProtobufTimeUtils.toJavaDuration(activityTask.getScheduleToCloseTimeout());
-      Optional<Duration> expiration =
-          timeout.compareTo(Duration.ZERO) > 0 ? Optional.of(timeout) : Optional.empty();
-      if (retryOptions.shouldRethrow(
-          result.getTaskFailed().getFailure(), expiration, attempt, elapsedTotal, sleepMillis)) {
-        return result;
-      } else {
-        result.setBackoff(Duration.ofMillis(sleepMillis));
+      long sleepMillis = retryOptions.calculateSleepTime(timingOutAttempt);
+
+      if (RetryOptionsUtils.isDeadlineReached(
+          executionContext.getScheduleToCloseDeadline(), sleepMillis)) {
+        executionContext.callback(
+            failed(
+                activityId,
+                RetryState.RETRY_STATE_TIMEOUT,
+                // scheduleToClose timeout is effectively replacing the original startToClose
+                newTimeoutFailure(
+                    TimeoutType.TIMEOUT_TYPE_SCHEDULE_TO_CLOSE, executionContext.getLastFailure()),
+                timingOutAttempt,
+                null));
+        return;
       }
 
-      // For small backoff we do local retry. Otherwise we will schedule timer on server side.
-      // TODO(maxim): Use timer queue for retries to avoid tying up a thread.
-      if (elapsedTask + sleepMillis < localRetryThresholdMs) {
-        try {
-          Thread.sleep(sleepMillis);
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          // if the activity thread got interrupted, we use the last failed result and exit fast
-          return result;
-        }
-        activityTask.setAttempt(attempt + 1);
-        return handleLocalActivity(
-            activityTask, activityStartTimeMs, localRetryThresholdMs, metricsScope);
-      } else {
-        return result;
+      if (executionContext.getLocalRetryDeadline().timeRemaining(TimeUnit.MILLISECONDS)
+          <= sleepMillis) {
+        executionContext.callback(
+            failed(
+                activityId,
+                LOCAL_RETRY_LIMIT_RETRY_STATE,
```

Comment:
Is this the state where we schedule a timer?

Reply:
Where we WILL schedule a timer in one of the upcoming PRs, yeah.